File: packages/ingest/src/parsers/cursor-agent.ts
import { readFileSync } from "node:fs";
import type { ParsedSession, ParsedTurn, ToolUseRecord, ToolResultRecord, SourcePlatform } from "@obs/shared";

/**
 * Cursor Agent transcript format (written by the benchmark adapter):
 *
 * Line 1: { type: "session_meta", sessionId, timestamp, platform, promptId, cwd, model }
 * Line 2: { type: "user", sessionId, timestamp, message: { role: "user", content: "..." } }
 * Line 3: { type: "assistant", sessionId, timestamp, message: { role: "assistant", content: [...] } }
 */

interface ContentBlock {
  type: string;
  text?: string;
  name?: string;
  input?: Record<string, unknown>;
  id?: string;
}

interface CursorLine {
  type: string;
  sessionId?: string;
  timestamp?: string;
  platform?: string;
  promptId?: string;
  cwd?: string;
  model?: string;
  message?: {
    role?: string;
    content?: string | ContentBlock[];
  };
}

/**
 * Parse a Cursor Agent JSONL transcript file into sessions.
 */
export function parseCursorAgentFile(filePath: string): ParsedSession[] {
  const content = readFileSync(filePath, "utf-8");
  const lines = content.split("\n").filter(Boolean);

  let sessionId: string | null = null;
  let cwd: string | null = null;
  let modelId: string | null = null;
  let startedAt: string | null = null;
  let endedAt: string | null = null;
  const turns: ParsedTurn[] = [];

  for (const line of lines) {
    let parsed: CursorLine;
    try {
      parsed = JSON.parse(line) as CursorLine;
    } catch {
      continue;
    }

    const ts = parsed.timestamp ?? new Date().toISOString();

    if (parsed.type === "session_meta") {
      sessionId = parsed.sessionId ?? null;
      cwd = parsed.cwd ?? null;
      modelId = parsed.model ?? null;
      if (!startedAt) startedAt = ts;
      continue;
    }

    if (!startedAt || ts < startedAt) startedAt = ts;
    if (!endedAt || ts > endedAt) endedAt = ts;

    if (parsed.sessionId && !sessionId) sessionId = parsed.sessionId;

    if (parsed.type === "user") {
      const turn = parseUserTurn(parsed, ts);
      if (turn) turns.push(turn);
    } else if (parsed.type === "assistant") {
      const turn = parseAssistantTurn(parsed, ts);
      if (turn) turns.push(turn);
    }
  }

  if (!sessionId || turns.length === 0) return [];

  return [{
    id: sessionId,
    platform: "cursor" as SourcePlatform,
    modelId,
    cwd,
    gitBranch: "__obs_bench__",
    startedAt: startedAt ?? new Date().toISOString(),
    endedAt,
    turns,
    filePath,
  }];
}

function parseUserTurn(line: CursorLine, timestamp: string): ParsedTurn | null {
  const content = line.message?.content;
  if (!content) return null;

  const textContent = typeof content === "string"
    ? content
    : Array.isArray(content)
      ? content.filter((b) => b.type === "text" && b.text).map((b) => b.text!).join("\n")
      : "";

  return {
    role: "user",
    textContent: textContent.trim(),
    toolUses: [],
    toolResults: [] as ToolResultRecord[],
    timestamp,
  };
}

function parseAssistantTurn(line: CursorLine, timestamp: string): ParsedTurn | null {
  const content = line.message?.content;
  if (!content) return null;

  let textContent = "";
  const toolUses: ToolUseRecord[] = [];

  if (typeof content === "string") {
    textContent = content;
  } else if (Array.isArray(content)) {
    for (const block of content) {
      if (block.type === "text" && block.text) {
        textContent += block.text + "\n";
      } else if (block.type === "tool_use") {
        toolUses.push({
          toolName: block.name ?? "unknown",
          input: (block.input as Record<string, unknown>) ?? {},
          id: block.id ?? "",
        });
      }
    }
  }

  return {
    role: "assistant",
    textContent: textContent.trim(),
    toolUses,
    toolResults: [],
    timestamp,
  };
}

File: data/benchmark/src/benchmark/parsers/cursor_agent.py
"""Cursor Agent JSONL transcript parser.

Port of packages/ingest/src/parsers/cursor-agent.ts
"""

from __future__ import annotations

import json

from .types import ParsedSession, ParsedTurn, ToolUseRecord


def parse_cursor_agent_file(file_path: str) -> list[ParsedSession]:
    """Parse a Cursor Agent JSONL transcript file into sessions."""
    with open(file_path, encoding="utf-8") as f:
        raw = f.read()
    return parse_cursor_agent_jsonl(raw, file_path)


def parse_cursor_agent_jsonl(content: str, file_path: str = "") -> list[ParsedSession]:
    """Parse Cursor Agent JSONL content string into sessions."""
    session_id: str | None = None
    cwd: str | None = None
    model_id: str | None = None
    started_at: str | None = None
    ended_at: str | None = None
    turns: list[ParsedTurn] = []

    for line in content.splitlines():
        if not line.strip():
            continue
        try:
            parsed = json.loads(line)
        except json.JSONDecodeError:
            continue

        ts = parsed.get("timestamp", "")
        line_type = parsed.get("type")

        if line_type == "session_meta":
            session_id = parsed.get("sessionId")
            cwd = parsed.get("cwd")
            model_id = parsed.get("model")
            if not started_at:
                started_at = ts
            continue

        if ts:
            if not started_at or ts < started_at:
                started_at = ts
            if not ended_at or ts > ended_at:
                ended_at = ts

        if parsed.get("sessionId") and not session_id:
            session_id = parsed["sessionId"]

        if line_type == "user":
            turn = _parse_user_turn(parsed, ts)
            if turn:
                turns.append(turn)
        elif line_type == "assistant":
            turn = _parse_assistant_turn(parsed, ts)
            if turn:
                turns.append(turn)

    if not session_id or not turns:
        return []

    from datetime import datetime, timezone

    return [
        ParsedSession(
            id=session_id,
            platform="cursor",
            model_id=model_id,
            cwd=cwd,
            git_branch="__obs_bench__",
            started_at=started_at or datetime.now(timezone.utc).isoformat(),
            ended_at=ended_at,
            turns=turns,
            file_path=file_path,
        )
    ]


def _parse_user_turn(line: dict, timestamp: str) -> ParsedTurn | None:
    msg = line.get("message")
    if not isinstance(msg, dict):
        return None
    content = msg.get("content")
    if content is None:
        return None

    if isinstance(content, str):
        text_content = content
    elif isinstance(content, list):
        text_content = "\n".join(
            b.get("text", "") for b in content if isinstance(b, dict) and b.get("type") == "text" and b.get("text")
        )
    else:
        text_content = ""

    return ParsedTurn(role="user", text_content=text_content.strip(), timestamp=timestamp)


def _parse_assistant_turn(line: dict, timestamp: str) -> ParsedTurn | None:
    msg = line.get("message")
    if not isinstance(msg, dict):
        return None
    content = msg.get("content")
    if content is None:
        return None

    text_content = ""
    tool_uses: list[ToolUseRecord] = []

    if isinstance(content, str):
        text_content = content
    elif isinstance(content, list):
        for block in content:
            if not isinstance(block, dict):
                continue
            if block.get("type") == "text" and block.get("text"):
                text_content += block["text"] + "\n"
            elif block.get("type") == "tool_use":
                tool_uses.append(
                    ToolUseRecord(
                        tool_name=block.get("name", "unknown"),
                        input=block.get("input", {}),
                        id=block.get("id", ""),
                    )
                )

    return ParsedTurn(
        role="assistant",
        text_content=text_content.strip(),
        tool_uses=tool_uses,
        tool_results=[],
        timestamp=timestamp,
    )
